skimage/morphology/_skeletonize_3d_cy.pyx.in

```diff
@@ -0,0 +1,430 @@
+    those point which can be removed without changing local connectivity in the
+    3x3x3 neighborhood of a point.
+
+    This routine implements the two-pass algorithm of [Lee94]_. Namely,
+    for each of the six border types (positive and negative x-, y- and z-),
+    the algorithm first collects all possibly deletable points, and then
+    performs a sequential rechecking.
+
+    The input, `img`, is assumed to be a 3D binary image in the
+    (p, r, c) format [i.e., C ordered array], filled by zeros (background) and
+    ones. Furthermore, `img` is assumed to be padded by zeros from all
+    directions --- this way the zero boundary conditions are automatic
+    and there is need to guard against out-of-bounds access.
+
+    """
+    cdef:
+        int unchanged_borders = 0, curr_border, num_borders
+        int borders[6]
+        npy_intp p, r, c
+        bint no_change
+
+        list simple_border_points
+        Py_ssize_t num_border_points, i, j
+        (npy_intp, npy_intp, npy_intp) point
+
+        pixel_type neighb[27]
+    borders[:] = [4, 3, 2, 1, 5, 6]
+
+    # no need to worry about the z direction if the original image is 2D.
+    if img.shape[0] == 3:
+        num_borders = 4
+    else:
+        num_borders = 6
+
+    # loop through the image several times until there is no change for all
+    # the six border types
+    while unchanged_borders < num_borders:
+        unchanged_borders = 0
+        for j in range(num_borders):
+            curr_border = borders[j]
+
+            simple_border_points = _loop_through(img, curr_border)
+
+            # sequential re-checking to preserve connectivity when deleting
+            # in a parallel way
+            no_change = True
+            num_border_points = len(simple_border_points)
+            for i in range(num_border_points):
+                point = simple_border_points[i]
+                p, r, c = point
+                get_neighborhood(img, p, r, c, neighb)
+                if is_simple_point(neighb):
+                    img[p, r, c] = 0
+                    no_change = False
+
+            if no_change:
+                unchanged_borders += 1
+
+    return np.asarray(img)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef list _loop_through(pixel_type[:, :, ::1] img,
+                        int curr_border):
+    """Inner loop of compute_thin_image.
+
+    The algorithm of [Lee94]_ proceeds in two steps: (1) six directions are
+    checked for simple border points to remove, and (2) these candidates are
+    sequentially rechecked, see Sec 3 of [Lee94]_ for rationale and discussion.
+
+    This routine implements the first step above: it loops over the image
+    for a given direction and assembles candidates for removal.
+
+    """
+    # This routine looks like it could be nogil, but actually it cannot be,
+    # because of `simple_border_points` being a python list which is being
+    # mutated.
```

Comment:
Could this be fixed by using e.g. 3 C++ vectors? That might be a significant performance boost. (Happy to leave this for a separate PR though.)